File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    private int firstPos(int[] nums, int target) {
        int l = 0, r = nums.length-1;
        while(l<=r) {
            int mid = l+((r-l)/2);
            if(nums[mid] == target) {
                if(mid>0 && nums[mid-1] != target) {
                    return mid;
                } else if(mid == 0) {
                    return mid;
                } else {
                    r = mid-1;
                }
            } else if(nums[mid] > target) {
                r = mid-1;
            } else {
                l = mid+1;
            }
        }
        return -1;
    }
    
    private int lastPos(int[] nums, int target) {
        int len = nums.length;
        int l = 0, r = len-1;
        while(l<=r) {
            int mid = l+((r-l)/2);
            if(nums[mid] == target) {
                if(mid<len-1 && nums[mid+1] != target) {
                    return mid;
                } else if(mid == len-1) {
                    return mid;
                } else {
                    l = mid+1;
                }
            } else if(nums[mid] > target) {
                r = mid-1;
            } else {
                l = mid+1;
            }
        }
        return -1; //---
    }
    public int[] searchRange(int[] nums, int target) {
        // int[] arr = new int[]{1,2,2,2,2,2,2,2,3};
        
        int ind1 = firstPos(nums, target);
        int ind2 = lastPos(nums, target);
        // System.out.println(ind1);
        // System.out.println(ind2);
        
        return new int[]{ind1,ind2};
    }
}

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.ts
const getStartInd = (nums: number[], target: number) => {
    let res = -1
    let l=0, r=nums.length-1

    while(l <= r) {
        const m = l + Math.trunc((r-l)/2)
        if(nums[m] === target) {
            res = m
            r = m-1
        } else if(nums[m] > target) {
            r = m-1
        } else {
            l = m+1
        }
    }
    return res
}

const getEndInd = (nums: number[], target: number) => {
    let res = -1
    let l=0, r=nums.length-1

    while(l <= r) {
        const m = l + Math.trunc((r-l)/2)
        if(nums[m] === target) {
            res = m
            l = m+1
        } else if(nums[m] > target) {
            r = m-1
        } else {
            l = m+1
        }
    }
    return res
}

function searchRange(nums: number[], target: number): number[] {
    const low = getStartInd(nums, target)
    const high = getEndInd(nums, target)
    // console.log(low, " : ", high)
    return [low, high]
};